src/System.Private.CoreLib/shared/System/Runtime/InteropServices/SafeHandle.cs

```diff
@@ -0,0 +1,258 @@
+                    ThrowHelper.ThrowObjectDisposedException(nameof(SafeHandle), ExceptionResource.ObjectDisposed_SafeHandleClosed);
+                }
+
+                // Not closed, let's propose an update (to the ref count, just add
+                // StateBits.RefCountOne to the state to effectively add 1 to the ref count).
+                // Continue doing this until the update succeeds (because nobody
+                // modifies the state field between the read and write operations) or
+                // the state moves to closed.
+                newState = oldState + StateBits.RefCountOne;
+            } while (Interlocked.CompareExchange(ref _state, newState, oldState) != oldState);
+            
+            // If we got here we managed to update the ref count while the state
+            // remained non closed. So we're done.
+            success = true;
+        }
+
+        public void DangerousRelease() => InternalRelease(disposeOrFinalizeOperation: false);
+
+        private void InternalRelease(bool disposeOrFinalizeOperation)
+        {
+            Debug.Assert(_fullyInitialized || disposeOrFinalizeOperation);
+
+            // See AddRef above for the design of the synchronization here. Basically we
+            // will try to decrement the current ref count and, if that would take us to
+            // zero refs, set the closed state on the handle as well.
+            bool performRelease = false;
+
+            // Might have to perform the following steps multiple times due to
+            // interference from other AddRef's and Release's.
+            int oldState, newState;
+            do
+            {
+                // First step is to read the current handle state. We use this cached
+                // value to predicate any modification we might decide to make to the
+                // state).
+                oldState = _state;
+
+                // If this is a Dispose operation we have additional requirements (to
+                // ensure that Dispose happens at most once as the comments in AddRef
+                // detail). We must check that the dispose bit is not set in the old
+                // state and, in the case of successful state update, leave the disposed
+                // bit set. Silently do nothing if Dispose has already been called.
+                if (disposeOrFinalizeOperation && ((oldState & StateBits.Disposed) != 0))
+                {
+                    return;
+                }
+
+                // We should never see a ref count of zero (that would imply we have
+                // unbalanced AddRef and Releases). (We might see a closed state before
+                // hitting zero though -- that can happen if SetHandleAsInvalid is
+                // used).
+                if ((oldState & StateBits.RefCount) == 0)
+                {
+                    ThrowHelper.ThrowObjectDisposedException(nameof(SafeHandle), ExceptionResource.ObjectDisposed_SafeHandleClosed);
```

Comment:
Does the throw helper help anything here? This is not a generic code, so there is not much code to save.

Reply:
Probably not.  We just already had a ThrowObjectDisposedException on ThrowHelper, I figured I'd use it.  I can change it back to a throw, though.

Reply:
> This is not a generic code, so there is not much code to save.

As an aside: Is using a throw helper only beneficial for generic types (also why is it beneficial for generics)? What about when the exception message string is built (for instance when using a format string with string interpolation)? We don't have that here, but was wondering regardless.

Reply:
> also why is it beneficial for generics

To avoid duplication of the generated code for each generic instantiation.

Reply:
Another situation where ThrowHelper is beneficial are small (inlineable) methods. It is not the case here either.